File: frontend/src/lib/taskUrgency.ts
/**
 * Единый источник правды для срочности задач — зеркало backend/app/services/task_urgency.py.
 *
 * due_date приходит с бэка как календарная дата (YYYY-MM-DD), без времени. Просрочка
 * считается в целых днях с полуночи после дедлайна: due_date === today — это ещё не
 * просрочка. Три вида задач (StudentTask, RoadmapTask, workspaceApi) используют эту
 * же функцию — иначе цвета разъедутся по экранам (см. ОС 30/07, Блок B).
 *
 * Пороги (Прил. № 3, п. 3.4): жёлтый < 24ч · оранжевый 24–48ч · красный 48–72ч ·
 * critical > 72ч (существенное нарушение — основание расторгнуть договор).
 */

export type Urgency = 'none' | 'yellow' | 'orange' | 'red' | 'critical'

// Зеркало NO_URGENCY_STATUSES из backend/app/services/task_urgency.py, а тот
// собран из TERMINAL_STATUSES | PAUSED_STATUSES в task_sla.py. Раньше здесь
// стоял один 'done': отменённая и принятая задачи считались горящими.
// Ждущая подписи регламента задача тоже не горит — часы SLA на неё не капают,
// штрафа за неё нет, и цвет на экране не должен обещать обратное.
// Тест на бэке пришпилен к точному составу набора: разойдётся — упадёт там.
const NO_URGENCY_STATUSES = new Set([
  'done',
  'accepted',
  'cancelled',
  'awaiting_signature',
  'blocked_by_agreement',
])

/**
 * Задача ещё требует работы. Дополнение к NO_URGENCY_STATUSES, поэтому «живая»
 * и «горит» считаются от одного набора и разойтись не могут.
 *
 * Нужна там, где список фильтруется на клиенте: фильтр status === 'open' прятал
 * ровно то, за что штрафуют, — просроченную (её фоновый цикл переводит в
 * overdue) и взятую в работу (in_progress).
 */
export function isTaskLive(status: string): boolean {
  return !NO_URGENCY_STATUSES.has(status)
}

function parseDateOnly(value: string): Date {
  const [year, month, day] = value.split('-').map(Number)
  return new Date(year, (month ?? 1) - 1, day ?? 1)
}

export function taskUrgency(dueDate: string | null | undefined, status: string, today: Date = new Date()): Urgency {
  if (!dueDate || NO_URGENCY_STATUSES.has(status)) return 'none'

  const due = parseDateOnly(dueDate)
  const reference = new Date(today.getFullYear(), today.getMonth(), today.getDate())
  const msPerDay = 24 * 60 * 60 * 1000
  const overdueDays = Math.round((reference.getTime() - due.getTime()) / msPerDay)

  if (overdueDays <= 0) return 'none'
  if (overdueDays <= 1) return 'yellow'
  if (overdueDays <= 2) return 'orange'
  if (overdueDays <= 3) return 'red'
  return 'critical'
}

File: backend/app/services/task_urgency.py
"""Единый источник правды для срочности задач — зеркало frontend/src/lib/taskUrgency.ts.

Срочность считается от due_date (дата, не время): просрочка в целых днях с полуночи
после дедлайна. Три вида задач (StudentTask, RoadmapTask, workspaceApi) используют
эту же функцию — иначе цвета разъедутся по экранам (см. ОС 30/07, Блок B).

Пороги (Прил. № 3, п. 3.4): 🟡 < 24ч · 🟠 24–48ч · 🔴 48–72ч · ⚫ > 72ч (critical,
существенное нарушение — основание расторгнуть договор).
"""
from datetime import date
from typing import Literal

from app.services.task_sla import PAUSED_STATUSES, TERMINAL_STATUSES

Urgency = Literal["none", "yellow", "orange", "red", "critical"]

# Статусы, в которых срочности нет вовсе. Раньше здесь стоял один "done", и
# отменённая или принятая задача считалась горящей — не было видно только
# потому, что эндпоинты фильтровали по status == open и до сюда её не пускали.
#
# Набор берётся из task_sla, а не переписывается литералами: срочность на
# экране обязана совпадать с тем, за что реально штрафуют. Отсюда же и пауза —
# ждущая подписи регламента задача не вина исполнителя, часы SLA на неё не
# капают, и красным её красить не за что.
#
# Значения, а не enum: сюда приходят и RoadmapTask (planned/in_progress/done),
# и StudentTask — общий знаменатель у них только строка.
NO_URGENCY_STATUSES = {status.value for status in TERMINAL_STATUSES | PAUSED_STATUSES}


def task_urgency(due_date: date | None, status: str, *, today: date | None = None) -> Urgency:
    """Срочность задачи по дедлайну и статусу.

    due_date — дата, а не datetime: просрочка отсчитывается от полуночи после
    дедлайна, то есть due_date == today ещё не просрочка (0 полных дней просрочки).
    """
    if due_date is None or status in NO_URGENCY_STATUSES:
        return "none"

    reference = today or date.today()
    overdue_days = (reference - due_date).days

    if overdue_days <= 0:
        return "none"
    if overdue_days <= 1:
        return "yellow"
    if overdue_days <= 2:
        return "orange"
    if overdue_days <= 3:
        return "red"
    return "critical"
